python/cudf/cudf/core/index.py

```diff
@@ -2673,6 +2676,139 @@ def categories(self):
+            right_col = bin_edges[1:]
+        else:
+            if freq and periods and end:
+                start = end - (freq * periods)
+            if freq and periods and start:
+                end = freq * periods + start
+            left_col = cupy.arange(start, end, freq)
+            if end is not None:
+                end = end + 1
+            if start is not None and freq is not None:
+                new_freq = start + freq
+            elif start is not None and freq is None:
+                new_freq = start
+            right_col = cupy.arange(new_freq, end, freq)
+            if len(left_col) != len(right_col):
+                if end and freq is not None:
+                    end = end - freq
+                left_col = cupy.arange(start, end, freq)
+            if len(right_col) == 0 or len(left_col) == 0:
+                return pd.IntervalIndex([], closed=closed)
+
+        interval_col = column.build_interval_column(
+            left_col, right_col, closed=closed
+        )
+        return cast(interval_range, IntervalIndex(interval_col))
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    IntervalIndex.
+
+    closed : {‘left’, ‘right’, ‘both’, ‘neither’}, default ‘right’
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    dtype : dtype or None, default None
+    If None, dtype will be inferred.
+
+    copy : bool, default False
+    Copy the input data.
+
+    name : object, optional
+    Name to be stored in the index.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy(deep=True)
+        out = Frame.__new__(cls)
+        kwargs = _setdefault_name(data, name=name)
+        if isinstance(data, IntervalColumn):
+            data = data
+        elif isinstance(data, pd.Series) and (is_interval_dtype(data.dtype)):
+            data = column.as_column(pa.Array.from_pandas(data), dtype=dtype,)
+        elif isinstance(data, (pd._libs.interval.Interval, pd.IntervalIndex)):
+            data = column.as_column(data, dtype=dtype,)
+        else:
+            if data is not None and data[0].closed != closed:
+                # when closed is not the same as the data's closed
+                # we need to change the data closed value,
+                # not sure how to do this on the gpu?
+                data = pd.arrays.IntervalArray(data, closed=closed)
+            data = column.as_column(
+                data, dtype="interval" if dtype is None else dtype
+            )
+            # dtype has already been taken care
+            dtype = None
+
+        out._initialize(data, **kwargs)
```

Comment:
Ooof, I'm trying to finish up `cut` using `cudf.IntervalIndex` and getting an error :/ 
Pandas allows us to access the intervals in IntervalIndex but we are currently getting an error with cudf when doing this. Not sure how to fix this. I'll try to experiment with it next week, but open to suggestions on how to address this!

```
In [3]: pd.IntervalIndex([pd.Interval(0,7),pd.Interval(2,8)])
Out[3]: 
IntervalIndex([(0, 7], (2, 8]],
              closed='right',
              dtype='interval[int64]')

In [4]: cudf.IntervalIndex([pd.Interval(0,7),pd.Interval(2,8)])
Out[4]: 
IntervalIndex([(0, 7], (2, 8]],
              closed='right',
              dtype='interval')

In [5]: p = pd.IntervalIndex([pd.Interval(0,7),pd.Interval(2,8)])

In [6]: c= cudf.IntervalIndex([pd.Interval(0,7),pd.Interval(2,8)])

In [7]: p[0]
Out[7]: Interval(0, 7, closed='right')

In [8]: c[0]
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-24-71463270cd6c> in <module>
----> 1 c[0]

~/cudf/python/cudf/cudf/core/index.py in __getitem__(self, index)
   2036 
   2037     def __getitem__(self, index):
-> 2038         res = self._values[index]
   2039         if not isinstance(index, int):
   2040             res = as_index(res)

~/cudf/python/cudf/cudf/core/column/column.py in __getitem__(self, arg)
    634     def __getitem__(self, arg) -> Union[ScalarLike, ColumnBase]:
    635         if is_scalar(arg):
--> 636             return self.element_indexing(int(arg))
    637         elif isinstance(arg, slice):
    638             start, stop, stride = arg.indices(len(self))

~/cudf/python/cudf/cudf/core/column/column.py in element_indexing(self, index)
    612             raise IndexError("single positional indexer is out-of-bounds")
    613 
--> 614         return libcudf.copying.get_element(self, idx).value
    615 
    616     def slice(self, start: int, stop: int, stride: int = None) -> ColumnBase:

~/cudf/python/cudf/cudf/_lib/copying.pyx in cudf._lib.copying.get_element()

RuntimeError: cuDF failure at: ../../../../src/copying/get_element.cu:166: get_element_functor not supported for struct_view
```